### PR TITLE
Android Toolset: implement animation extraction from HOMM2.GOG

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
If `HOMM2.GOG` is present in a ZIP archive with HoMM2 assets. On request from people writing reviews on Google Play.

ISO9660 module was mostly taken from [stephenc/java-iso-tools](https://github.com/stephenc/java-iso-tools) repo (LGPL license, currently unmaintained). I removed a lot of unnecessary code that did nothing useful for us but require unnecessary dependencies (including natural monsters like [Apache Hadoop](https://github.com/stephenc/java-iso-tools/blob/334caa435f58ebaa8a5f69eefc2e5ac247c0d6c2/loop-fs-spi/pom.xml#L50-L54)), fixed a few bugs (for instance, the original code treated files with the `Hidden` attribute in the ISO image as directories and caused a crash), modernized the original code a bit, formatted it according to our standards and added our copyright strings above the original ones.

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , since we are greatly expanding Java code I think we should enable Java code analysis within SonarQube. I don't remember that we enabled it. What do you think?

File: android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java [hunk 1]
```diff
@@ -92,6 +121,58 @@ static boolean extractHoMM2AssetsFromZip( final File externalFilesDir, final Inp
...
+        // It is allowed to extract only files located in these subdirectories
+        final Set<String> allowedSubdirNames = new HashSet<>();
+        allowedSubdirNames.add( "anim" );
+
+        final Set<File> allowedSubdirs = new HashSet<>();
+        for ( String name : allowedSubdirNames ) {
+            allowedSubdirs.add( new File( externalFilesDir, name ).getCanonicalFile() );
+        }
+
+        boolean result = false;
```
Comment (ihhub) [inline android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:137]: Can we please rename it into something like **containsValidFile**? **result** is a very generic name :)

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I left just few comments here. Would you mind check them?

File: android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java [hunk 2]
```diff
@@ -92,6 +121,58 @@ static boolean extractHoMM2AssetsFromZip( final File externalFilesDir, final Inp
         return result;
     }
 
+    // Returns true if at least one animation was found and extracted, otherwise returns false
+    @SuppressWarnings( "IOStreamConstructor" ) // Files.newOutputStream() requires API level 26
+    private static boolean extractAnimationsFromISO( final File externalFilesDir, final File isoFile ) throws IOException
+    {
+        // It is allowed to extract only files located in these subdirectories
+        final Set<String> allowedSubdirNames = new HashSet<>();
+        allowedSubdirNames.add( "anim" );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:130]: @ihhub The logic is the same as in `extractHoMM2AssetsFromZip()`: `getHoMM2AssetSubpath()` is called that contains case-insensitive logic (all path elements are lowercased via `String.toLowerCase()` before comparison).

File: android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java [hunk 3]
```diff
@@ -125,4 +207,42 @@ private static boolean isValidHoMM2AssetPath( final File path, final Set<File> a
...
+        final int chunkSize = 2352;
+        final byte[] chunk = new byte[chunkSize];
+
+        int chunkOffset = 0;
+
+        while ( true ) {
+            final int bytesRead = gogStream.read( chunk, chunkOffset, chunkSize - chunkOffset );
+
+            // EOF
+            if ( bytesRead < 0 ) {
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:223]: @ihhub There is a [documented behavior](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[],%20int,%20int)): result of `read()` can be zero only if `len` is zero and this is impossible here.

File: android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java (hunk 1 above)
Comment (oleg-derevenetz) [reply android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:137]: Well, I often use the name `result` for a variable that will be used as a function return value (see the `extractHoMM2AssetsFromZip()` in this module, `extractAnimationsFromISO()` has absolutely the same logic). But if you insist, I can rename it :)

File: android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:223]: Got it! Thank you for the explanation!

**Comment by ihhub** [comment]
@oleg-derevenetz , thank you so much for this change!

**Pull request merged**